### ISSUE TITLE
A Convenience method for converting between data formats

### ISSUE THREAD

**Issue by b5** [qri-io/dataset#127]
We keep repeating data format conversion code all over the place. I think we should add a primitive for converting a dataset from one data format to the other:

```golang
package dsio

// ConvertAll consumes an io.Reader of data in the format of a given structure to dst structure
func ConvertAll(dst dataset.DataFormat, st *dataset.Structure, r io.Reader) ([]byte, error) {
  entr, err := NewReader(st, r)
  // if err != nil ...

  buf, err := NewEntryBuffer(&dataset.Structure{Format : dst, Schema: st.Schema})
  // if err != nil ...
  
  err := EachEntry(r, func(_ int, ent Entry, err error) error {
    if err != nil {
      return err
    }
    return buf.WriteEntry(ent)
  })
  // if err != nil ...

  return buf.Bytes(), nil
}
```

Thoughts?

**Comment by dustmop** [comment]
Are there more examples you could provide in the codebase of where this type of conversion happens? In https://github.com/qri-io/skytf/pull/1 the real goal is not to get from File -> []byte (raw bytes to be parsed by json.Unmarshal), but from File -> interface{} (map or array object of arbitrary structure, to be passed to skytf.lib.Marshal). It seems like sort of an anti-pattern to go from a stream of data -> blob-of-everything, because it's generally inefficient and greedy, but it just happens to be the right thing to do skylark because we need all the data available for the Transform's code.

It might end up what we really want in a general utility is io.Reader -> io.Reader, which will support cases were streaming is needed instead. Or maybe we actually do want io.Reader -> interface{}. Seeing where else this sort of thing happens would help inform that decision.

**Comment by b5** [comment]
Most obvious example is in `core/StructuredData`:

https://github.com/qri-io/qri/blob/8a9dc8dee4123be5e4828c9eef753ac3458135c0/core/datasets.go#L630-L661

The more I think about it, we may have managed to isolate this conversion process to that one spot. I'll try to remember others.

Totally agreed about `io.Reader` preservation and not falling down to `[]byte`. I think the key insight you've identified is the need to make the `EntryReader` a "bidirectional relationship" with `io.Reader`. Currently `dsio.EntryReader` "upgrades" an `io.Reader` with the implicit expectation that once you create a dsio.EntryReader, you'll only call `ReadEntry`, which manages calls to the wrapped reader for you.

Long term, I think we should consider some implementation of `Read(p []byte) (num i, err error)` from a `dsio.EntryReader`. If we piped together a `dsio.EntryReader` -> `dsio.EntryBuffer`, and had some way to `Read(p []byte) (int, error)` off that buffer, we can do streaming conversion that's totally opaque to the end caller of `Read`.

So maybe we park this for now & pick it up later

**Comment by b5** [comment]
closed by #128
